File: lib/p2p/chain/action/p2pchat.ts
import {ChainMessage, MsgType} from '@/libs/p2p/message';
import {BaseAction, PayloadType} from '@/libs/p2p/chain/baseaction';
import {dataBlockService} from '@/libs/p2p/chain/datablock';

/**
 在chain目录下的采用自定义protocol "/chain"的方式自己实现的功能
 */
export class P2pChatAction extends BaseAction {
  constructor(msgType: MsgType) {
    super(msgType);
  }

  async chat(connectPeerId: string, data: any, targetPeerId: string): Promise<any> {
    let chainMessage: ChainMessage = this.prepareSend(connectPeerId, data, targetPeerId);
    // 已经使用signal protocol加密，不用再加密
    //chainMessage.NeedEncrypt = true
    //chainMessage.NeedSlice = true
    let response: ChainMessage = await this.send(chainMessage);
    if (response) {
      return response;
    }

    return null;
  }

  async receive(chainMessage: ChainMessage): Promise<ChainMessage> {
    let _chainMessage = await super.receive(chainMessage);
    let srcPeerId: string = chainMessage.SrcPeerId;
    let payload: string;
    if (_chainMessage.PayloadType === PayloadType.DataBlock) {
      let _dataBlock = _chainMessage.Payload;
      await dataBlockService.decrypt(_dataBlock);
      payload = _dataBlock.payload;
    } else {
      payload = chainMessage.Payload;
    }
    if (chainMessage && p2pChatAction.receivers) {
      p2pChatAction.receivers.forEach(async (receiver, key) => {
        await receiver(srcPeerId, payload);
      });
      return null;
    }
  }
}

export let p2pChatAction = new P2pChatAction(MsgType.P2PCHAT);

File: lib/p2p/chain/action/p2pchat.dart
import 'package:colla_chat/entity/p2p/chain_message.dart';
import 'package:colla_chat/p2p/chain/action/chat.dart';
import 'package:colla_chat/p2p/chain/baseaction.dart';

///服务进行消息推送
class P2pChatAction extends ChatAction {
  P2pChatAction(super.msgType);

  @override
  Future<void> transferPayload(ChainMessage chainMessage) async {
    super.transferPayload(chainMessage);
    if (chainMessage.payloadType == PayloadType.dataBlock.name) {
      Map payload = {};
      var jsons = chainMessage.payload;
      if (jsons is List) {
        for (var json in jsons) {
          payload = json['payload'];
        }
      }
      chainMessage.payload = payload;
    }
  }
}

final p2pChatAction = P2pChatAction(MsgType.P2PCHAT);
